File: packages/engine/src/worker/runner/python/context.py
import hash_util


class AgentContext:
    def __init__(self, sim_ctx, ctx_batch, state_snapshot, i_agent_in_sim):
        # The context batch is sim-wide, so hide it from the user.
        # (The user should only see one current agent through an AgentContext object.)
        self.__sim_ctx = sim_ctx
        self.__getters = sim_ctx.getters  # Used often
        self.__cols = ctx_batch.cols
        self.state_snapshot = state_snapshot
        self.__idx_in_sim = i_agent_in_sim  # (As opposed to agent index in its group)

    def to_json(self):
        r = {}
        for field_name in self.__cols:
            r[field_name] = hash_util.json_deepcopy(self.__getattr__(field_name))
        return r

    def globals(self):
        return self.__sim_ctx.globals()

    def data(self):
        return self.__sim_ctx.data()

    def step(self):
        return self.__sim_ctx.step()

    def __getattr__(self, field_name):
        print(f"__getattr__ cols: {self.__dict__['_AgentContext__cols']}")
        elem = self.__dict__['_AgentContext__cols'][field_name][self.__dict__['_AgentContext__idx_in_sim']]
        getter = self.__dict__['_AgentContext__getters'].get(field_name)
        return elem if getter is None else getter(self, elem)

    # Context is immutable, so there's no `__setattr__`.


class GroupContext:
    def __init__(self, sim_ctx, ctx_batch, state_snapshot, group_start_idx):
        # The context batch is sim-wide, so hide it from the user.
        # (The user should only see one current group through a GroupContext object.)
        self.__sim_ctx = sim_ctx
        self.__ctx_batch = ctx_batch
        self.state_snapshot = state_snapshot
        self.__start_idx = group_start_idx

    def get_agent(self, i_agent_in_group, old_agent_ctx=None):
        idx_in_sim = i_agent_in_group + self.__start_idx
        if old_agent_ctx is not None:  # Reuse AgentContext object for performance.
            old_agent_ctx.__AgentContext_idx_in_sim = idx_in_sim
            return old_agent_ctx

        return AgentContext(
            self.__sim_ctx, self.__ctx_batch, self.state_snapshot, idx_in_sim
        )

    def globals(self):
        return self.__sim_ctx.globals()

    def data(self):
        return self.__sim_ctx.data()

    def step(self):
        return self.__sim_ctx.step()


class Snapshot:
    def __init__(self, agent_pool, message_pool):
        self.agent_pool = agent_pool
        self.message_pool = message_pool


class SimContext:
    def __init__(self, getters, experiment_ctx, sim_globals):
        self.getters = getters
        self.__step = 0
        self.__experiment_ctx = experiment_ctx
        self.__globals = sim_globals
        self.__ctx_batch = None
        self.state_snapshot = Snapshot(None, None)

    # Invalidates existing `GroupContext` and `AgentContext` objects.
    def set_batch(self, ctx_batch):
        self.__ctx_batch = ctx_batch
        # TODO: group state agent index --> sim context agent index

    # Invalidates existing `GroupContext` and `AgentContext` objects.
    def set_snapshot(self, agent_pool, message_pool):
        self.state_snapshot.agent_pool = agent_pool
        self.state_snapshot.message_pool = message_pool

    def set_step(self, cur_step):
        self.__step = cur_step

    # TODO: step getter method

    def get_group(self, i_group):
        return GroupContext(
            self,
            self.__ctx_batch,
            self.state_snapshot,
            i_group
        )

    def get_agent(self, i_agent_in_sim, old_agent_ctx=None):
        if old_agent_ctx is not None:  # Reuse AgentContext object for performance.
            old_agent_ctx.__AgentContext_idx_in_sim = i_agent_in_sim
            return old_agent_ctx

        return AgentContext(
            self, self.__ctx_batch, self.state_snapshot, i_agent_in_sim
        )

    def globals(self):
        return self.__globals

    def data(self):
        return self.__experiment_ctx.data()

    def step(self):
        return self.__step


class SimInitContext:
    def __init__(self, experiment_ctx, sim_globals, agent_schema):
        self.__experiment_ctx = experiment_ctx
        self.__globals = sim_globals  # TODO: Freeze somehow
        self.agent_schema = agent_schema

    def globals(self):
        return self.__globals

    def data(self):
        return self.__experiment_ctx.data()


class ExperimentContext:
    def __init__(self, datasets):
        self.__datasets = datasets  # TODO: Freeze somehow

    def data(self):
        return self.__datasets

File: packages/engine/src/worker/runner/javascript/context.js
/// `x` must not contain reference cycles.
const deepfreeze = (x) => {
  for (var k in x) {
    if (x.hasOwnProperty(k)) {
      const v = x[k];
      if (typeof v === "object") deepfreeze(v);
    }
  }
  return Object.freeze(x);
};

const gen_agent_ctx_getter = (name, getter) => {
  return getter
    ? function () {
        return getter(this, this.__cols[name][this.__idx_in_sim]);
      }
    : function () {
        return this.__cols[name][this.__idx_in_sim];
      };
};

// TODO: agent view? (__sim_ctx field, this.__cols = sim_ctx.ctx_batch.cols,
//       defineProperty(..., "state_snapshot", ... __sim_ctx.state_snapshot))
const gen_agent_ctx = (ctx_schema, getters) => {
  const AgentContext = function (
    ctx_batch,
    state_snapshot,
    current_step,
    i_agent_in_sim,
    globals,
    experiment_ctx,
  ) {
    // The context batch is sim-wide, so hide it from the user.
    // (The user should only see one current agent through an AgentContext object.)
    this.__experiment_ctx = experiment_ctx;
    this.__ctx_batch = ctx_batch;
    this.__cols = ctx_batch.cols; // Used often
    this.__current_step = current_step;
    this.__idx_in_sim = i_agent_in_sim; // (As opposed to agent index in its group)
    this.__globals = globals;
    this.state_snapshot = state_snapshot;
  };

  AgentContext.prototype.data = function () {
    return this.__experiment_ctx.data();
  };

  AgentContext.prototype.globals = function () {
    return this.__globals;
  };

  AgentContext.prototype.step = function () {
    return this.__current_step;
  };

  for (var i_field = 0; i_field < ctx_schema.fields.length; ++i_field) {
    // `name` is the name of some context batch column
    const name = ctx_schema.fields[i_field].name;
    const getter = gen_agent_ctx_getter(name, getters[name]);
    AgentContext.prototype[name] = getter;
  }

  return Object.seal(AgentContext);
};

const gen_group_ctx = (AgentContext) => {
  const GroupContext = function (
    ctx_batch,
    state_snapshot,
    current_step,
    group_start_idx,
    globals,
    experiment_ctx,
  ) {
    // The context batch is sim-wide, so hide it from the user.
    // (The user should only see one current group through a GroupContext object.)
    this.__experiment_ctx = experiment_ctx;
    this.__globals = globals;
    this.__ctx_batch = ctx_batch;
    this.__start_idx = group_start_idx;
    this.__current_step = current_step;
    this.state_snapshot = state_snapshot;
  };

  GroupContext.prototype.data = function () {
    return this.__experiment_ctx.data();
  };

  GroupContext.prototype.globals = function () {
    return this.__globals;
  };

  GroupContext.prototype.step = function () {
    return this.__current_step;
  };

  GroupContext.prototype.get_agent = function (
    i_agent_in_group,
    old_agent_ctx,
  ) {
    const idx_in_sim = i_agent_in_group + this.__start_idx;
    if (old_agent_ctx) {
      // Reuse AgentContext object for performance.
      old_agent_ctx.__idx_in_sim = idx_in_sim;
      return old_agent_ctx;
    }
    return Object.seal(
      new AgentContext(
        this.__ctx_batch,
        this.state_snapshot,
        this.__current_step,
        idx_in_sim,
        this.__globals,
        this.__experiment_ctx,
      ),
    );
  };

  return Object.seal(GroupContext);
};

export const gen_sim_ctx = (ctx_schema, getters) => {
  const AgentContext = gen_agent_ctx(ctx_schema, getters);
  const GroupContext = gen_group_ctx(AgentContext);

  const SimContext = function (experiment_ctx, globals) {
    this.__experiment_ctx = experiment_ctx;
    this.__globals = deepfreeze(globals);
    this.__ctx_batch = null;
    this.__group_start_idxs = null;
    this.__current_step = null;
    this.state_snapshot = null;
  };

  SimContext.prototype.data = function () {
    return this.__experiment_ctx.data();
  };

  SimContext.prototype.globals = function () {
    return this.__globals;
  };

  SimContext.prototype.step = function () {
    return this.__current_step;
  };

  /// Invalidates existing `GroupContext` and `AgentContext` objects.
  SimContext.prototype.set_batch = function (
    ctx_batch,
    state_group_start_idxs,
    current_step,
  ) {
    this.__ctx_batch = ctx_batch;
    this.__group_start_idxs = state_group_start_idxs;
    this.__current_step = current_step;
  };

  /// Invalidates existing `GroupContext` and `AgentContext` objects.
  SimContext.prototype.sync_snapshot = function (state_snapshot) {
    this.state_snapshot = state_snapshot;
  };

  SimContext.prototype.get_group = function (i_group) {
    return Object.seal(
      new GroupContext(
        this.__ctx_batch,
        this.state_snapshot,
        this.__current_step,
        this.__group_start_idxs[i_group],
        this.__globals,
        this.__experiment_ctx,
      ),
    );
  };

  SimContext.prototype.get_agent = function (i_agent_in_sim, old_agent_ctx) {
    if (old_agent_ctx) {
      // Reuse AgentContext object for performance.
      old_agent_ctx.__idx_in_sim = i_agent_in_sim;
      return old_agent_ctx;
    }

    return Object.seal(
      new AgentContext(
        this.__ctx_batch,
        this.state_snapshot,
        this.__current_step,
        i_agent_in_sim,
        this.__globals,
        this.__experiment_ctx,
      ),
    );
  };

  return Object.seal(SimContext);
};

export const ExperimentContext = function (datasets) {
  this.__datasets = deepfreeze(datasets);
};

ExperimentContext.prototype.data = function () {
  return this.__datasets;
};

export const SimInitContext = function (experiment_ctx, globals, agent_schema) {
  this.__experiment_ctx = experiment_ctx;
  this.__globals = deepfreeze(globals);
  this.agent_schema = agent_schema;
};

SimInitContext.prototype.data = function () {
  return this.__experiment_ctx.data();
};

SimInitContext.prototype.globals = function () {
  return this.__globals;
};
